openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,173 +113,193 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
+	if err != nil && !os.IsNotExist(err) {
 		return err
 	}
+	if err == nil {
+		fmt.Fprintf(os.Stderr, "dataDir %s is present on %s\n", o.DataDir, o.TargetName)
+		dataDirExists = true
+	}
 
 	client, err := o.getClient()
-	if err != nil && memberDirExists {
-		// we weren't able to get client and need to return based previous memberDir so we can restart.  This is the off and on again flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get client, but memberDir %s is present on %s, err=%s. Returning with no error.\n", memberDir, o.TargetName, err)
-		fmt.Printf(o.TargetName)
+
+	// Condition: unknown membership but dataDir exists
+	// possible reasons for this condition.
+	// 1.) single node etcd cluster
+	// 2.) transient networking problem
+	// 3.) on and off flow
+	// Result: empty initial config
+	if err != nil && dataDirExists {
+		fmt.Fprintf(os.Stderr, "Couldn't get client, but dataDir %s is present on %s, err=%s. Returning with no error.\n", o.DataDir, o.TargetName, err)
 		return nil
-	} else if err != nil {
+	}
+	if err != nil {
 		return err
 	}
 	defer client.Close()
 
-	var targetMember *etcdserverpb.Member
-	var allMembers []*etcdserverpb.Member
+	target := url.URL{
+		Scheme: o.TargetPeerURLScheme,
+		Host:   fmt.Sprintf("%s:%s", o.TargetPeerURLHost, o.TargetPeerURLPort),
+	}
+
 	for i := 0; i < 10; i++ {
 		fmt.Fprintf(os.Stderr, "#### attempt %d\n", i)
-		targetMember, allMembers, err = o.checkForTarget(client)
 
-		for _, member := range allMembers {
-			fmt.Fprintf(os.Stderr, "      member=%v\n", stringifyMember(member))
+		// check member list on each iteration for changes
+		cluster, err := client.MemberList(context.TODO())
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "member list request failed: %v", err)
+			continue
+		}
+		logCurrentMembership(cluster.Members)
+		targetMember, found := checkTargetMember(target, cluster.Members)
+
+		// Condition: unstarted member found
+		// This member is part of the cluster but has not yet started. We know this because the name is populated at
+		// runtime which this member does not have.
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
-		fmt.Fprintf(os.Stderr, "      target=%v, err=%v\n", stringifyMember(targetMember), err)
 
-		// we're done because we found what we want.
-		if targetMember != nil && err == nil {
-			break
+		// Condition: unstarted member found with dataDir
+		// This member is part of the cluster but has not yet started yet has a dataDir
+		// Result: archive old dataDir and return Error
+		if found && targetMember.Name == "" && dataDirExists {
+			archivedDir, err := archiveDataDir(o.DataDir)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("member %q is unstarted but dataDir exists. %q has been archived to %q", target.String(), o.DataDir, archivedDir)
```

Comment:
This code will never reach, because the condition on L175 matches before the condition on L183.

Reply:
Just need to swap the `if` blocks to put this one ahead.

Reply:
right `! dataDirExists` should be added to 175 conditions will resolve.